src/target/sam4l.c

```diff
@@ -0,0 +1,410 @@
+			DEBUG("SAM4L: target_check_error returned true\n");
+		}
+		return true;
+	}
+	return false;
+}
+
+/*
+ * We've been reset, make sure we take the core out of reset
+ */
+void
+sam4l_extended_reset(target *t)
+{
+	uint32_t	reg;
+	int i;
+
+	DEBUG("SAM4L: Extended Reset\n");
+	/* enable SMAP in case we're dealing with a non-TCK SRST */
+	target_mem_write32(t, SMAP_CR, 0x1); /* enable SMAP */
+	reg = target_mem_read32(t, SMAP_SR);
+	DEBUG("\nSAM4L: SMAP_SR has 0x%08lx\n", (long unsigned int) reg);
+	if ((reg & SMAP_SR_HCR) != 0) {
+		/* write '1' bit to the status clear register */
+		target_mem_write32(t, SMAP_SCR, SMAP_SR_HCR);
+		/* waiting 250 loops for it to reset is arbitrary, it should happen right away */
+		for (i = 0; i < 250; i++) {
+			reg = target_mem_read32(t, SMAP_SR);
+		}
+		/* not sure what to do if we can't reset that bit */
+		if (i > 249) {
+			DEBUG("\nSAM4L: Reset failed. SMAP_SR has 0x%08lx\n", (long unsigned int) reg);
+		}
+	}
+	/* reset bus error if for some reason SMAP was disabled */
+	target_check_error(t);
+}
+
+/*
+ * sam4l_flash_command
+ *
+ * Helper function, wait for the flash controller to be ready to receive a
+ * command. Then send it the command, page number, and the authorization
+ * key (always 0xA5) in the command register.
+ *
+ * Need the target struct to call the mem_read32 and mem_write32 function
+ * pointers.
+ */
+static int
+sam4l_flash_command(target *t, uint32_t page, uint32_t cmd)
+{
+	uint32_t cmd_reg;
+	uint32_t status;
+	int	timeout;
+	DEBUG("\nSAM4L: sam4l_flash_command: FSR: 0x%08x, page = %d, command = %d\n", 
+		(unsigned int)(FLASHCALW_FSR), (int) page, (int) cmd);
+	/* wait for Flash controller ready */
+	for (timeout = 0; timeout < FLASH_TIMEOUT; timeout++) {
+		status = target_mem_read32(t, FLASHCALW_FSR);
+		if (status & FLASHCALW_FSR_FRDY) {
+			break;
+		}
+	}
+	if (timeout == FLASH_TIMEOUT) {
+		DEBUG("\nSAM4L: sam4l_flash_command: Not ready! Status = 0x%08x\n", (unsigned int) status);
+		return -1; /* Failed */
+	}
+	/* load up the new command */
+	cmd_reg = (cmd & FLASHCALW_FCMD_CMD_MASK) |
+			  ((page & FLASHCALW_FCMD_PAGEN_MASK) << FLASHCALW_FCMD_PAGEN_SHIFT) |
+		  	  (0xA5 << FLASHCALW_FCMD_KEY_SHIFT);
+	DEBUG("\nSAM4L: sam4l_flash_command: Wrting command word 0x%08x\n", (unsigned int) cmd_reg);
+	/* and kick it off */
+	target_mem_write32(t, FLASHCALW_FCMD, cmd_reg);
+	/* don't actually wait for it to finish, the next command will stall if it is not done */
+	return 0;
+}
+
+/*
+ * Write data from 'src' into flash using the algorithim provided by
+ * Atmel in their data sheet.
+ */
+static int
+sam4l_flash_write_buf(struct target_flash *f, target_addr addr, const void *src, size_t len)
+{
+	target *t = f->t;
+	uint32_t *src_data = (uint32_t *)src;
+	uint32_t ndx;
+	uint16_t page;
+	
+	DEBUG("\nSAM4L: sam4l_flash_write_buf: addr = 0x%08lx, len %d\n", (long unsigned int) addr, (int) len);
+	/* This will fail with unaligned writes, the write_buf version */
+	while (len) {
```

Comment:
This loop isn't needed. The buffered write will always be called with exactly one sector.